src/librustc_mir/interpret/intrinsics.rs

```diff
@@ -122,6 +122,56 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
                     self.binop_with_overflow(bin_op, lhs, rhs, dest)?;
                 }
             }
+            "saturating_add" => {
+                let l = self.read_immediate(args[0])?;
+                let r = self.read_immediate(args[1])?;
+                let (val, overflowed) = self.binary_op_imm(BinOp::Add, l, r)?;
+                if overflowed {
+                    let first_term: u128 = l.to_scalar()?.to_bits(l.layout.size)?;
+                    let num_bits = l.layout.size.bits();
+                    let val = if l.layout.abi.is_signed() {
+                        // For signed addition the saturated value depends on the
```

Comment:
isn't this missing a situation like `5i8 + (-200i8)`? We are not looking at the rhs at all (in contrast to what the old non-intrinsic code was doing)

Reply:
@oli-obk I was assuming that an earlier compiler stage (parser?) would catch invalid tokens (200i8). Just ran a small test and it gives "error: literal out of range for i8"

Reply:
Oh, that was stupid of me, ok, checking the lhs works just as well as checking the rhs. I'll check the logic again with this info

Reply:
Please add a comment explaining this reasoning.